library/core/src/alloc/mod.rs

```diff
@@ -94,6 +94,18 @@ pub unsafe trait AllocRef {
     /// The returned block may have a larger size than specified by `layout.size()`, and may or may
     /// not have its contents initialized.
     ///
+    /// Note that you may not rely on this method actually getting called, even if there are calls
+    /// to it in the source. The optimizer may detect allocation/deallocation pairs that it can
+    /// instead move to stack allocations/deallocations and thus never invoke the allocator here.
+    /// More concretely, the following code example is unsound, irrespective of whether your
+    /// custom allocator allows counting how many allocations have happened.
+    ///
+    /// ```rust,ignore
+    /// drop(Box::new(0));
+    /// let number_of_heap_allocs = /* call private allocator API */;
+    /// unsafe { std::intrinsics::assume(number_of_heap_allocs == 0); }
```

Comment:
If the allocator has an API to query the number of allocs, I think the compiler is not allowed to bypass this, as calling `alloc` has a side effect (e.g. incrementing a counter in the allocator).

```suggestion
    /// let mut number_of_heap_allocs = 0;
    /// drop(Box::new(0));
    /// number_of_heap_allocs += 1;
```

The use must not rely on `number_of_heap_allocs` then.

Reply:
> If the allocator has an API to query the number of allocs, I think the compiler is not allowed to bypass this, as calling alloc has a side effect (e.g. incrementing a counter in the allocator).

At least for the global allocator, that would be a massive optimization regression compared to `malloc`.

Reply:
Ah, I didn't know that LLVM might handle the allocation calls specially.

Reply:
Note that Rust allocation calls currently go through `__rust_alloc`, which is not handled specially by LLVM. But if we ever *do* want to treat them specially, we need to carefully document this.

Or rather, this is the extend of the current special treatment:
```
; Function Attrs: nounwind nonlazybind uwtable
declare noalias i8* @__rust_alloc(i64, i64) unnamed_addr #0
```
Notably is `nounwind` and `noalias` -- so this must not unwind, and the returned pointer must be "fresh", i.e., not alias any other pointer that the calling context might use. If that is not already the case, this should also be documented.

Reply:
uh, this is slightly off topic, but the `AllocRef::alloc` docs specifically state that we are allowed to panic: https://doc.rust-lang.org/std/alloc/trait.AllocRef.html#errors, so unless the `__rust_alloc` body generated by `#[global_allocator]` catches panics, we do have a problem.

Reply:
Yeah, this is a problem that I brought up before:
* https://github.com/rust-lang/rust/pull/77738#discussion_r502300229
* https://github.com/rust-lang/rust/issues/66741#issuecomment-706079813

It's a bit tricky because *panicking* is okay as long as we have `-C panic=abort`.

Reply:
They do receive a special treatment from LLVM as distributed along with rustc. See commit titled "Add knowledge of __rust\_{alloc,realloc,dealloc}".

Reply:
> Note that Rust allocation calls currently go through __rust_alloc, which is not handled specially by LLVM (malloc is a "magic name"). But if we ever do want to treat them specially, we need to carefully document this.

Actually our LLVM fork does treat `__rust_alloc`, `__rust_realloc` and `__rust_dealloc` specially: https://github.com/llvm/llvm-project/commit/4e6611b7b04e1125139ba384fef2c96b5a63e614

Reply:
Oh, thanks for chiming in! That is good to know, and it makes comments like this even more important.